File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.kt
class Solution {
    fun findDuplicate(arr: IntArray): Int {
        var slow=arr[0]
        var fast=arr[arr[0]]
        while(slow!=fast){
            slow=arr[slow]
            fast=arr[arr[fast]]
        }
        slow=0
        while(slow!=fast){
            slow=arr[slow]
            fast=arr[fast]
        }
        return slow
    }
}

File: 0287-find-the-duplicate-number/0287-find-the-duplicate-number.cpp
class Solution {
public:
    int findDuplicate(vector<int>& arr) {
        if(arr.size()<=2)return arr[0];
        int slow=arr[0],fast=arr[arr[0]];
        while(slow!=fast){
            slow=arr[slow];
            fast=arr[arr[fast]];
        }
        slow=0;
        while(slow!=fast){
            slow=arr[slow];
            fast=arr[fast];
        }
        return slow;
    }
};